stmhal/sdcard.c

```diff
@@ -85,6 +86,29 @@ void sdcard_init(void) {
     // invalidate the sd_handle
     sd_handle.Instance = NULL;
 
+    // configure the SD card detect pin
+    // we do this here so we can detect if the SD card is inserted before powering it on
+    GPIO_Init_Structure.Mode = GPIO_MODE_INPUT;
+    GPIO_Init_Structure.Speed = GPIO_SPEED_LOW;
+#if defined(MICROPY_HW_SDCARD_DETECT_PIN)
+    // assume a mechanical switch is used
+    GPIO_Init_Structure.Pull = MICROPY_HW_SDCARD_DETECT_PULL;
+    GPIO_Init_Structure.Pin = MICROPY_HW_SDCARD_DETECT_PIN.pin_mask;
+    HAL_GPIO_Init(MICROPY_HW_SDCARD_DETECT_PIN.gpio, &GPIO_Init_Structure);
+#elif defined(MICROPY_HW_SDCARD_DETECT_PULL)
+    // assume no mechanical switch, so assume CD/DAT3 is used for card detect.
+    GPIO_Init_Structure.Pull = MICROPY_HW_SDCARD_DETECT_PULL;
+    GPIO_Init_Structure.Pin = pin_C11.pin_mask;
+    HAL_GPIO_Init(pin_C11.gpio, &GPIO_Init_Structure);
+    // capture sdcard present state before the IO pin is remapped to SDIO functionality
+    sdcard_present = (HAL_GPIO_ReadPin(pin_C11.gpio, pin_C11.pin_mask) == MICROPY_HW_SDCARD_DETECT_PRESENT);
+    // TODO: do we need to issue ACMD42 (SD_CMD_SD_APP_SET_CLR_CARD_DETECT) to disable the internal pullup resistor?
+#else
+    // SD Card Detect not defined.  Just assume it's there.
```

Comment:
Unfortunately, there doesn't seem to be any way of detecting if the card is present without trying to send a command to the card. I agree its an oversight, but I still have to make it work. So for teensy (which isn't using shared code with stmhal for sdcard), If the equivalent of HAL_SD_Init fails, then I assume that the card is missing. I cache that and have sdcard_is_present() return the cached value.

The stmhal could use card-detected or CD/DAT3, but I can't use either technique on teensy. Since there isn't currently an stmhal based board which doesn't use some type of card-detect we don't need to change anything yet. But there may be a board down the road.